### PR TITLE
Fix optional groups sometimes messing up non-empty match groups

### PR DESCRIPTION
**Pull request by petee-d**
Description in https://github.com/alecthomas/participle/issues/148

### CODE REVIEW

File: nodes.go [hunk 1]
```diff
@@ -216,17 +216,17 @@ func (g *group) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.Va
 	for ; matches < max; matches++ {
 		branch := ctx.Branch()
 		v, err := g.expr.Parse(branch, parent)
-		out = append(out, v...)
 		if err != nil {
 			ctx.MaybeUpdateError(err)
 			// Optional part failed to match.
 			if ctx.Stop(err, branch) {
+				out = append(out, v...) // Try to return as much of the parse tree as possible
```
Comment (petee-d) [inline nodes.go:223]: I'm not entirely sure about this line, but my understanding is that if the error should terminate matching the entire group, the unsuccessfully attempted match output should be propagated. Can you check my thinking here? There wasn't a test in the code base that would catch whether this is done or not and I couldn't think of any.

Comment (alecthomas) [reply nodes.go:223]: It sounds like you're saying it should be propagated (which it should) but you've removed it? 

It should remain though, yes.

Comment (petee-d) [reply nodes.go:223]: No, I didn't remove it. This particular line keeps propagating the partial output for errors that abort the entire group (it appends it to the output before returning it along with the error). I just wasn't sure whether in this situation the partial output would be anything useful for the user, but it made more sense to me to keep it.

Comment (petee-d) [reply nodes.go:223]: BTW, the link to the issue where I explained the bug and this change in the description of this PR could have easily been missed, so just to be sure, here is the context: https://github.com/alecthomas/participle/issues/148

Comment (alecthomas) [reply nodes.go:223]: Ah I see, sorry for some reason the your original comment was a few lines up on mobile, but here it makes sense. LGTM

**Pull request merged**